controllers/appmesh/cloudmap_controller.go

```diff
@@ -0,0 +1,888 @@
+			return obj.(*cloudmapServiceCacheItem).key, nil
+		}, 60*time.Second),
+		enqueueRequestsForPodEvents: cloudmap.NewEnqueueRequestsForPodEvents(k8sClient, log),
+	}
+}
+
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+func (r *cloudMapReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	return runtime.HandleReconcileError(r.reconcile(req), r.log)
+}
+
+func (r *cloudMapReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appmesh.VirtualNode{}).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, r.enqueueRequestsForPodEvents).
+		Complete(r)
+}
+
+func (r *cloudMapReconciler) reconcile(req ctrl.Request) error {
+	ctx := context.Background()
+	log := r.log.WithValues("Virtualnode", req.NamespacedName)
+
+	vNode := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vNode); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("VNode: ", "awsName: ", vNode.Spec.AWSName,
+		"vNode.PodSelector: ", vNode.Spec.PodSelector)
+	if !vNode.DeletionTimestamp.IsZero() {
+		return r.deleteCloudMapResources(ctx, vNode)
+	}
+	return r.reconcileVirtualNodeWithCloudMap(ctx, vNode)
+}
+
+func (r *cloudMapReconciler) reconcileVirtualNodeWithCloudMap(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	if !r.isCloudMapEnabledForVirtualNode(vNode) {
+		r.log.V(1).Info("CloudMap not enabled for", " virtualNode: %s", vNode.Name)
+		return nil
+	}
+
+	if !r.isPodSelectorDefinedForVirtualNode(vNode) {
+		r.log.V(1).Info("PodSelector not defined for", "virtualNode: %s", vNode.Name)
+		return nil
+	}
+
+	cloudMapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	creatorRequestID := vNode.UID
+
+	if err := r.finalizerManager.AddFinalizers(ctx, vNode, k8s.FinalizerAWSCloudMapResources); err != nil {
+		r.log.Error(err, "..while adding cloudmap resource finalizer")
+		return err
+	}
+
+	_, err := r.getOrCreateCloudMapService(ctx, cloudMapConfig, string(creatorRequestID))
+	if err != nil {
+		r.log.Error(err, "failed to create cloudMap Service for", "vNode: ", vNode.Name)
+		return err
+	}
+
+	readyPods, notReadyPods, ignoredPods, perr := r.virtualNodeEndPointResolver.ResolveCloudMapEndPoints(ctx, vNode)
+	if perr != nil {
+		r.log.Error(perr, "failed to get pods for", "vNode: ", vNode.Name)
+		return perr
+	}
+
+	r.log.V(1).Info("EndPoints: ", "Ready Pods Count: ", len(readyPods.Items),
+		"Unready Pods Count: ", len(notReadyPods.Items), "Ignored Pods Count: ", len(ignoredPods.Items))
+	if len(readyPods.Items) == 0 && len(notReadyPods.Items) == 0 {
+		if len(ignoredPods.Items) > 0 {
+			//Handle the scenario where the deployment behind the Virtualnode is deleted
+			if err := r.deleteAWSCloudMapServiceInstances(ctx, cloudMapConfig); err != nil {
```

Comment:
we should just always update cloudmap if `readyPods+notReadypods` don't match our cached cloudmap status.
K8s can miss events, so it's possible to only get events when all pods are gone. (including ignorePods)

Reply:
will change it